src/identifier.rs

```diff
@@ -70,18 +70,95 @@ use crate::alloc::alloc::{alloc, dealloc, Layout};
 #[cfg(no_from_ne_bytes)]
 use crate::backport::FromNeBytes;
 use core::mem;
-use core::num::{NonZeroU64, NonZeroUsize};
+use core::num::NonZeroUsize;
 use core::ptr;
+use core::ptr::NonNull;
 use core::slice;
 use core::str;
 
 #[repr(transparent)]
 pub(crate) struct Identifier {
-    repr: NonZeroU64,
+    repr: Repr,
+}
+
+use crate::identifier::repr::Repr;
+
+mod repr {
+    use core::ptr::NonNull;
+
+    #[cfg(target_pointer_width = "64")]
+    #[derive(PartialEq, Eq, Clone, Copy)]
+    #[repr(C)]
+    pub struct Repr {
+        ptr: NonNull<u8>,
+    }
+
+    #[cfg(target_pointer_width = "32")]
+    #[derive(PartialEq, Eq, Clone, Copy)]
+    #[repr(C)]
+    pub struct Repr {
+        ptr: NonNull<u8>,
+        remainder: usize,
+    }
+
+    impl Repr {
+        #[cfg(target_pointer_width = "32")]
+        pub fn as_u64(self) -> u64 {
+            self.ptr.as_ptr() as u64 + ((self.remainder as u64) << 32)
+        }
+
+        #[cfg(target_pointer_width = "64")]
+        pub fn as_u64(self) -> u64 {
+            self.ptr.as_ptr() as u64
+        }
+
+        pub fn as_ptr(self) -> *const u8 {
+            self.ptr.as_ptr()
+        }
+
+        #[cfg(target_pointer_width = "32")]
+        pub const fn from_ptr(ptr: NonNull<u8>) -> Self {
+            Repr {
+                ptr: ptr,
+                remainder: 0,
+            }
+        }
+
+        #[cfg(target_pointer_width = "32")]
+        pub unsafe fn from_inline(bytes: u64) -> Repr {
+            Repr {
+                ptr: unsafe { NonNull::new_unchecked(bytes as usize as *mut u8) },
```

Comment:
I think this assumes the target is little endian. On big endian, the bytes at the lower address in memory are more significant, so the less significant half of `bytes` would be holding the *end* of the inline string, which would all be 0 bytes for strings length 4 or shorter -- see the call site here:

https://github.com/dtolnay/semver/blob/0b381703bb5873da58867b9746b7598552bb658d/src/identifier.rs#L174-L180

Reply:
`--target mips64-unknown-linux-gnuabi64` works pretty well for little-endian testing.